### PR TITLE
A few simplifications of `maybe_changed_since`

### PR DESCRIPTION
**Pull request by michalt**
As observed in #188, the method looks very similar to the combination of `probe` and `validate_memoized_value`, but is different enough that I didn't see to completely replace it. The core difference is what the name suggest ;) -- it really is focused on answering the question of whether something changed, whereas the other methods are focused on providing/validating an actual value.  Do let me know if I'm missing something though!

For reviewing, consider looking at the commits one by one -- I tried to separate the change to formatting/location from the ones affecting the behavior.

### CODE REVIEW

File: src/derived/slot.rs [hunk 1]
```diff
@@ -429,55 +429,29 @@ where
             self, revision, revision_now,
         );
 
-        // Acquire read lock to start. In some of the arms below, we
-        // drop this explicitly.
-        let state = self.state.read();
+        // Check the current state and potentially block if there's a computation in progress.
+        let state = match self.probe(db, self.state.read(), runtime, revision_now) {
```
Comment (nikomatsakis) [inline src/derived/slot.rs:433]: So the problem here is that we are going to clone the value in this case, which isn't necessary

File: src/derived/slot.rs [hunk 2]
```diff
@@ -522,53 +522,51 @@ where
...
+        // the interim. =) Therefore, we have to probe the current `self.state`  again and in some
+        // cases we ought to do nothing.
+        let mut state = self.state.write();
+        match &mut *state {
+            QueryState::Memoized(memo) => {
+                if memo.verified_at == revision_now {
+                    // Since we started verifying inputs, somebody else has come along and updated
+                    // this value (they may even have recomputed it). Therefore, we should not touch
+                    // this memo.
+                    return memo.changed_at > revision_since;
```
Comment (nikomatsakis) [inline src/derived/slot.rs:545]: This change seems correct though.

**Comment by nikomatsakis** [comment]
Sorry for taking forever to review this PR -- overall I like the idea of simplifying that function, but I don't think introducing a clone of the values is the right way to go. That said, I'm wanting to kind of refactor and simplify this code-path anyway, so maybe it's a moot point...

File: src/derived/slot.rs (hunk 1 above)
Comment (michalt) [reply src/derived/slot.rs:433]: Indeed! Thanks for catching this.

**Comment by michalt** [comment]
Thanks for taking a look! Since you're planning to simplify this code anyway, I'll just close the PR.

**Pull request closed without merging**